File: client/Assets/Ronriku/Scripts/Domain/Shop/Shop.cs
using System;
using System.Collections.Generic;
using Ronriku.Domain.Daily;
using Ronriku.Domain.Figures;
using Ronriku.Domain.Player;

namespace Ronriku.Domain.Shop
{
    public sealed class ShopItem
    {
        public string Id { get; }
        public ulong Seed { get; }
        public int Size { get; }
        public Figure Figure { get; }
        /// <summary>Shard price, or -1 when the item is SOL-only.</summary>
        public int PriceShards { get; }

        public ShopItem(ulong seed, int size)
        {
            Seed = seed;
            Size = size;
            Id = $"fig-{seed}-{size}";
            Figure = FigureGenerator.Generate(seed, size);
            PriceShards = Economy.FigurePrice(Figure.Rarity);
        }

        public bool SolOnly => PriceShards < 0;
    }

    public enum PurchaseResult { Ok, AlreadyOwned, InsufficientShards, SolOnly }

    /// <summary>Six figures rotate every UTC day; everyone sees the same shelf.</summary>
    public static class ShopCatalogue
    {
        public const int ShelfSize = 6;
        private static readonly int[] Sizes = { 3, 3, 4, 4, 5, 5 };

        public static IReadOnlyList<ShopItem> ForDay(int day)
        {
            long daySeed = DailyPlan.Mix(DailyCalendar.Seed(day), 0x53484F50);
            var items = new List<ShopItem>(ShelfSize);
            for (int i = 0; i < ShelfSize; i++)
                items.Add(new ShopItem(unchecked((ulong)DailyPlan.Mix(daySeed, i)), Sizes[i]));
            return items;
        }

        public static PurchaseResult Buy(PlayerProfile profile, ShopItem item)
        {
            if (profile == null) throw new ArgumentNullException(nameof(profile));
            foreach (OwnedFigure f in profile.figures) if (f.id == item.Id) return PurchaseResult.AlreadyOwned;
            if (item.SolOnly) return PurchaseResult.SolOnly;
            if (profile.shards < item.PriceShards) return PurchaseResult.InsufficientShards;
            profile.shards -= item.PriceShards;
            profile.figures.Add(new OwnedFigure { id = item.Id, seed = item.Seed.ToString(), size = item.Size, acquired = "shop" });
            return PurchaseResult.Ok;
        }

        public static Figure Build(OwnedFigure owned) =>
            FigureGenerator.Generate(ulong.TryParse(owned.seed, out ulong s) ? s : 1UL, owned.size);
    }
}

File: packages/core/src/shop.ts
/**
 * Port of Ronriku.Domain.Shop.ShopCatalogue (C#): six figures rotate every UTC day and everyone sees
 * the same shelf.
 *   daySeed = Mix(DailyCalendar.Seed(day), 0x53484F50)          ("SHOP")
 *   item i  = FigureGenerator.Generate((ulong)Mix(daySeed, i), [3,3,4,4,5,5][i])
 *   id      = "fig-{seed as ulong decimal}-{size}", price by rarity (Legendary = SOL only)
 */
import { dailySeed, mix } from "./daily/index.js";
import { figurePrice } from "./economy.js";
import type { Figure } from "./figures/figure.js";
import { generateFigure } from "./figures/generator.js";
import { u64 } from "./rng.js";

export const SHELF_SIZE = 6;
const SHELF_SIZES = [3, 3, 4, 4, 5, 5];

export interface ShopShelfItem {
  id: string;
  /** Unsigned 64-bit generator seed. */
  seed: bigint;
  size: number;
  figure: Figure;
  /** Shard price, or -1 when SOL-only. */
  priceShards: number;
}

export function shopItem(seed: bigint, size: number): ShopShelfItem {
  const s = u64(seed);
  const figure = generateFigure(s, size);
  return { id: `fig-${s}-${size}`, seed: s, size, figure, priceShards: figurePrice(figure.rarity) };
}

export function shopShelf(day: number): ShopShelfItem[] {
  const daySeed = mix(dailySeed(day), 0x53484f50n);
  return SHELF_SIZES.map((size, i) => shopItem(mix(daySeed, BigInt(i)), size));
}
